### ISSUE TITLE
Better support for `bytea` type

### ISSUE THREAD

**Issue by jakob** [jakob/Postico#128]
Postico currently has only limited support for `bytea` columns. Postico can display images stored in bytea columns in the sidebar, but apart from that there is no way to look at the data.

I've just discussed some ideas with @apeiros, who uses bytea columns to store textual data that may contain invalid code points. He suggested that the data should be displayed directly in the table view, in one of the following formats:

1. ascii with escape sequences, eg. `hello\x00w\xc3\xb6rld`
2. hex, eg. `0x68656c6c6f0077c3b6726c64`
3. utf-8, eg. `hello\x00wörld`

It would be nice if Postico did some auto-detection to display data in a sensible way (eg. don't try to display images as text), and also allow the user to view data in a specific way.

I'd love to hear feedback on this. Do people actually use the bytea type? And what kind of data do you store inside?

**Comment by jakob** [comment]
I've just received a suggestion from a Postico user via email:

> It would be very nice if there were 

> 1. an option to view bytea field contents encoded as hex or base64
> 2. the export and import function would work for those fields (especially copy-paste), because right now, exporting will just render those fields as empty values

**Comment by mtwilliams** [comment]
Any movement on this? I'm looking to use a `bytea` column to store Erlang terms in [ETF](http://erlang.org/doc/apps/erts/erl_ext_dist.html).

**Comment by jakob** [comment]
@mtwilliams no updates yet -- as you can see there isn't a lot of interest on this issue.

It looks like ETF is a complex binary format -- would displaying that in Postico even make sense? What form would be most useful, I assume just the raw bytes in hex format?

**Comment by mtwilliams** [comment]
The raw bytes in hex would work well. Especially if it came with a pop-out window a la HxD.

**Comment by mcculley** [comment]
I have a table with a column that uses the bytea type. The data is images or short strings encoded as UTF-8. I would really like for Postico to be able to show me the strings.

**Comment by tsutsu** [comment]
People who need deterministic primary keys tend to use binary-encoded UUIDs in a `bytea`-typed primary-key column. Framework's such as Elixir's ORM Ecto [have first-class support for this approach](https://blog.fourk.io/uuids-as-primary-keys-in-phoenix-with-ecto-and-elixir-1dd79e1ecc2e).

Personally, I'm using `bytea` for a number of things related to cryptocurrency, namely columns (which are often primary- or foreign-keys) that hold binary-encodings of SHA256 or RIPEMD160 hashes (those being the ID formats of most crypto-ledgers' blocks and addresses, respectively.)

Essentially, in both of these cases, `bytea` is only being used as a lower-overhead equivalent to storing the hex representation of the value directly in a `text` column.

So, in both of the above cases, being able to see/edit the hex representation of the `bytea` would be perfect.

**Comment by jakob** [comment]
@tsutsu I agree that displaying bytea as HEX would be useful for hashes, but as far as I understand the blog post you linked to talks about the PostgreSQL UUID type, which Postico does support properly.

**Comment by banteg** [comment]
I'm using `bytea` to store hashes, an option to display them as hex would be very useful.

**Comment by DR-BoneZ** [comment]
I am using `bytea` to store bitcoin locking scripts and being able to copy/paste them as hex would be very useful.

**Comment by ralphcallaway** [comment]
i use the bytea for storing several hashes and would love to be able to get the value in postico table explorer

**Comment by Fyb3roptik** [comment]
+1 for me. I store hashed items in my database and being able to verify the hash with what I show in the log would be great!

**Comment by voikya** [comment]
One more vote here for hash storage. I'd love to see the data displayed in hex format.

**Comment by supreethdevm** [comment]
one more vote for displaying bytea as HEX

**Comment by tomdcc** [comment]
Another vote. For my purposes hex viewing doesn't matter, but being able to export the binary data somehow would be very useful. Just having Copy Cell work so that I can then use pbpaste to put the data somewhere would be nice.

**Comment by pscott31** [comment]
another vote on this; our database is chock full of 32 byte bytea hashes that I would like to view, copy, paste etc.. as hex strings. I'll buy a license if you add this as I love the program otherwise!

**Comment by jakob** [comment]
After only 7 years of deliberation, I've started working on this feature. I've decided on supporting at least the following display modes:

1. HEX (by far the most popular demand)
2. Base 64 (useful for reducing size of exports)
3. Crockfords Base 32 (useful for ULIDs, recently suggested by a customer via email)

Possibly I'll also add these modes later (slightly harder because they are variable length, so I can't compute them on the fly and will need some kind of caching to avoid performance problems)

4. ASCII with octal escapes (this is how PostgreSQL traditionally displayed bytea, could be useful for data that is partially textual)
5. UTF-8 mode (this one is tricky, because I don't know how to best handle data that is not valid UTF-8)

Development of this feature is in the bytea-display branch which is available here: https://releases.eggerapps.at/postico2/changelog?branch_id=764

Here's my plan:

- [x] Implement global setting for bytea display mode
- [ ] Implement per-column setting

- [x] Format bytea values in the Table View
- [ ] Allow editing formatted bytea values
- [ ] Format bytea values in the sidebar

- [ ] Add filters for bytea columns:
  - [ ] is equal / is not equal
  - [ ] contains / does not contain (maybe only for HEX? would be confusing for Base32/Base64 since the search string would not be visible in the encoded text if it is not aligned to 5/6 byte groups)
  - [ ] maybe length < / = / > filters?

- [ ] Format bytea values in the SQL Query results
- [ ] Sorting SQL Query results

- [ ] Write test cases for encoding / decoding the various formats


If you have any more suggestions w/r/t bytea columns, let me know!

**Comment by mmaa** [comment]
Would `uuid` columns also support the option to display in Crockford's Base 32? I use `uuid` columns to store ULIDs as primary keys, and it would be nice not to have to convert between the two display formats. I.e., my app displays the primary keys in Crockford's Base 32, but Postico displays them as UUIDs.

**Comment by jakob** [comment]
I merged the bytea feature into the main branch. So it is now available in the latest development build, which you can download from the changelist: https://releases.eggerapps.at/postico2/changelog?update_channel=1

It's still a work in progress (see todo list above), but I think it's already useful in its current state.

Please let me know if you run into an issues!

**Comment by hakanderyal** [comment]
Hey Jakob, thanks for adding this feature (I'm the user who requested the Crockford Base32 via email).

There is a problem with Crockford Base32 conversion. (has been since the beginning). 

Test result for a value:
My ULID as C-Base32: `01HWSHT6Z85Y0HFJ2NEDH2W2PM`
Postico C-Base32: `067K678VX0QR25Y8ANSP4BGATG`

Tool for testing online: https://ulid.page

While the values doesn't match, the value from Postico is a valid Crockford Base32 value. I tested with online converters, but they give the same value as Postico. Checked the ULID spec, it specifies Crockford Base32 with no additional modifications. So I've no idea about the reasons of the discrepancy.

**Comment by jakob** [comment]
@hakanderyal Thanks for letting me know about the issue. Looks like ULIDs are aligned in a different way. I created a separate issue for this problem.

**Comment by jakob** [comment]
@hakanderyal I've fixed the Crockford / ULID alignment issue in Build 9698. You can download it from https://releases.eggerapps.at/postico2/changelog?update_channel=1

**Comment by hakanderyal** [comment]
Thanks 🙏  Just tested it, no problems.
